src/schedule/zephyr_dp_schedule_native.c

```diff
@@ -0,0 +1,572 @@
+
+	list_for_item(tlist, &dp_sch->tasks) {
+		curr_task = container_of(tlist, struct task, list);
+		pdata = curr_task->priv_data;
+		struct processing_module *mod = pdata->mod;
+		bool trigger_task = false;
+
+		/* decrease number of LL ticks/cycles left till the module reaches its deadline */
+		if (mod->dp_startup_delay && is_ll_post_run && pdata->ll_cycles_to_start) {
+			pdata->ll_cycles_to_start--;
+			if (!pdata->ll_cycles_to_start)
+				/* delayed start complete, clear startup delay flag.
+				 * see dp_startup_delay comment for details
+				 */
+				mod->dp_startup_delay = false;
+		}
+
+		if (curr_task->state == SOF_TASK_STATE_QUEUED &&
+		    mod->dev->state >= COMP_STATE_ACTIVE) {
+			/* trigger the task */
+			curr_task->state = SOF_TASK_STATE_RUNNING;
+			trigger_task = true;
+
+			pdata->pend_proc++;
+			k_sem_give(pdata->sem);
+		}
+		if (curr_task->state == SOF_TASK_STATE_RUNNING) {
+			/* (re) calculate deadline for all running tasks */
+			/* get module deadline in us */
+			uint32_t deadline = module_get_deadline(mod);
+
+			/* if a deadline cannot be calculated, use a fixed value relative to its
+			 * first start
+			 */
+			if (deadline >= UINT32_MAX / 2 && trigger_task)
+				deadline = module_get_lpt(mod);
+
+			if (deadline < UINT32_MAX) {
+				/* round down to 1ms */
+				deadline = deadline / 1000;
+
+				/* calculate number of ticks */
+				deadline = deadline * (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC / 1000);
+
+				/* add to "NOW", overflows are OK  */
+				deadline = dp_sch->last_ll_tick_timestamp + deadline;
+
+				/* set in Zephyr. Note that it may be in past, it does not matter,
+				 * Zephyr still will schedule the thread with earlier deadline
+				 * first
+				 */
+				k_thread_absolute_deadline_set(pdata->thread_id, deadline);
+			}
+		}
+	}
+}
+
+/* Thread function called in component context, on target core */
+void dp_thread_fn(void *p1, void *p2, void *p3)
+{
+	struct task *task = p1;
+	struct task_dp_pdata *task_pdata = task->priv_data;
+	struct processing_module *pmod = task_pdata->mod;
+	unsigned int lock_key;
+	enum task_state state;
+	bool first = true;
+	bool task_stop;
+
+	ARG_UNUSED(p2);
+	ARG_UNUSED(p3);
+
+	do {
+		if (first) {
+			/*
+			 * The IPC thread is waiting for the thread to be
+			 * started, it can proceed now.
+			 */
+			first = false;
+			k_sem_give(&dp_sync[cpu_get_id()]);
+		}
+
+		/*
+		 * The thread is started immediately after creation, it stops on
+		 * semaphore. Semaphore is signaled to handle IPC or process audio
+		 */
+		k_sem_take(task_pdata->sem, K_FOREVER);
+
+		lock_key = scheduler_dp_lock(task->core);
+
+		unsigned char pend_ipc = task_pdata->pend_ipc,
+			pend_proc = task_pdata->pend_proc;
+
+		task_pdata->pend_proc = 0;
+		task_pdata->pend_ipc = 0;
+
+		scheduler_dp_unlock(lock_key);
+
+		/* Only 0:1, 1:0 and 1:1 are valid */
+		if (pend_ipc > 1 || pend_proc > 1) {
+			tr_err(&dp_tr, "Invalid wake up %u:%u", pend_proc, pend_ipc);
+			continue;
+		}
```

Comment:
What you probably need here is `k_event_wait` or `k_event_wait_safe` along with `k_event_set ` on signalling side.

Reply:
maybe, will check, thanks